File: selection/_027/Main.java
package selection._027;

import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int y = sc.nextInt();
        int x = sc.nextInt();
        int[][] a = new int[x][y];
        for (int i = 0; i < x; i++) {
            for (int j = 0; j < y; j++) {
                a[i][j] = sc.nextInt();
            }
        }
        sc.close();

        int result = 0;
        for (int i = 0; i < x; i++) {
            for (int j = 0; j < y; j++) {
                if (a[i][j] == 0) {
                    continue;
                }
                int cur = dfs(i, j, 1, a, x, y);
                result = Math.max(result, cur);
            }
        }

        System.out.println(result);
    }

    static int dfs(int x, int y, int count, int[][] a, int maxX, int maxY) {
        a[x][y] = 0;
        int[] dx = new int[]{-1,0,1,0};
        int[] dy = new int[]{0,1,0,-1};

        int result = 0;
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (!(0 <= nx && nx < maxX && 0 <= ny && ny < maxY)) {
                continue;
            }
            if (a[nx][ny] == 0) {
                continue;
            }
            int cur = dfs(nx, ny, count + 1, a, maxX, maxY);
            result = Math.max(result, cur);
        }
        a[x][y] = 1;

        return Math.max(count, result);
    }
}

File: selection/_027/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var n, m int
	fmt.Scan(&n, &m)

	g := make([][]int, n)
	for i := 0; i < n; i++ {
		g[i] = make([]int, m)
		for j := 0; j < m; j++ {
			fmt.Scan(&g[i][j])
		}
	}

	result := 0
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if g[i][j] == 1 {
				dfs(i, j, 0, g, &result)
			}
		}
	}

	fmt.Println(result)
}

var dx = []int{0, -1, 0, 1}
var dy = []int{1, 0, -1, 0}

func dfs(x, y, cnt int, g [][]int, result *int) {
	cnt++
	*result = int(math.Max(float64(*result), float64(cnt)))
	g[x][y] = 0

	for i := 0; i < 4; i++ {
		nx, ny := dx[i]+x, dy[i]+y
		if 0 > nx || nx >= len(g) || 0 > ny || ny >= len(g[0]) {
			continue
		}
		if g[nx][ny] == 0 {
			continue
		}

		dfs(nx, ny, cnt, g, result)
	}

	g[x][y] = 1
}
